dali/imgcodec/parsers/bmp.cc

```diff
@@ -12,19 +12,150 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <stdexcept>
 #include "dali/imgcodec/parsers/bmp.h"
+#include "dali/core/byte_io.h"
+#include "dali/core/format.h"
+#include "dali/core/small_vector.h"
 
 namespace dali {
 namespace imgcodec {
 
+namespace  {
+
+enum BmpCompressionType {
+  BMP_COMPRESSION_RGB       = 0,
+  BMP_COMPRESSION_RLE8      = 1,
+  BMP_COMPRESSION_RLE4      = 2,
+  BMP_COMPRESSION_BITFIELDS = 3
+};
+
+bool is_color_palette(InputStream *stream, int ncolors, int palette_entry_size) {
+  SmallVector<uint8_t, 8> entry;
+  entry.resize(palette_entry_size);
+  for (int i = 0; i < ncolors; i++) {
+    stream->ReadAll(entry.data(), palette_entry_size);
+
+    const auto b = entry[0], g = entry[1], r = entry[2];  // a = p[3];
+    if (b != g || b != r)
+      return true;
+  }
+  return false;
+}
+
+int number_of_channels(InputStream *stream,
+                       int bpp, int compression_type, int ncolors = 0,
+                       int palette_entry_size = 0) {
+  if (compression_type == BMP_COMPRESSION_RGB || compression_type == BMP_COMPRESSION_RLE8) {
+    if (bpp <= 8 && ncolors <= (1_i64 << bpp)) {
+      return is_color_palette(stream, ncolors, palette_entry_size) ? 3 : 1;
+    } else if (bpp == 24) {
+      return 3;
+    } else if (bpp == 32) {
+      return 4;
+    }
+  } else if (compression_type == BMP_COMPRESSION_BITFIELDS) {
+    if (bpp == 16) {
+      return 3;
+    } else if (bpp == 32) {
+      return 4;
+    }
+  }
+
+  throw std::runtime_error(make_string(
+    "configuration not supported. bpp: ", bpp, " compression_type:", compression_type,
+    "ncolors:", ncolors));
+}
+
+}  // namespace
+
+struct BitmapCoreHeader {
+  uint32_t header_size;
+  uint16_t width, heigth, planes, bpp;
+};
+static_assert(sizeof(BitmapCoreHeader) == 12);
+
+struct BitmapInfoHeader {
+  int32_t header_size;
+  int32_t width, heigth;
+  uint16_t planes, bpp;
+  uint32_t compression, image_size;
+  int32_t x_pixels_per_meter, y_pixels_per_meter;
+  uint32_t colors_used, colors_important;
+};
+static_assert(sizeof(BitmapInfoHeader) == 40);
+
 ImageInfo BmpParser::GetInfo(ImageSource *encoded) const {
+  auto stream = encoded->Open();
+  ssize_t length = stream->Size();
+
+  DALI_ENFORCE(length >= 18);
+
+  static constexpr int kHeaderStart = 14;
+  stream->SeekRead(kHeaderStart);
+  uint32_t header_size = stream->ReadOne<uint32_t>();
+  stream->Skip<uint32_t>(-1);  // we'll read it again - it's part of the header struct
```

Comment:
took me a while to understand that Skip -1 means going back. Perhaps using SeekRead directly would read better.

Reply:
It was specifically designed this way - and even is documented:
```
  /**
   * @brief Skips `count` objects in the stream
   *
   * Skips over the length of `count` objects of given type (by default char,
   * because sizeof(char) == 1).
   *
   * NOTE: Negative skips are allowed.
   *
   * @tparam T type of the object to skip; defaults to `char`
   * @param count the number of objects to skip
   */
   ```

Reply:
@klecki 
Answering some questions:
> Btw, why the Codecs from the design are called Decoders, or am I confusing them?
We decided decoders was a better name, by request of the reviewers of a previous PR. The design document has been updated (yesterday).

> And lastly, maybe I'm nitpicking a bit, but I was assigned to 4 PRs on Monday morning, and the code seems quite modular: why do we have a PR that implements two distinct things, instead of two PRs with one thing each? Even the title with `+` kinda suggests it. That would be probably a good example :)
Perhaps we should extract the Bmp parser to a different PR.